index.go

```diff
@@ -126,6 +139,78 @@ func (idx *faissIndex) Add(x []float32) error {
 	return nil
 }
 
+func (idx *faissIndex) GetClusterAssignment() (map[int64][]int64, error) {
+
+	clusterVectorIDMap := make(map[int64][]int64)
+
+	ivfPtr := C.faiss_IndexIVF_cast(idx.cPtr())
+	if ivfPtr == nil {
+		return clusterVectorIDMap, nil
+	}
+
+	nlist := C.faiss_IndexIVF_nlist(idx.idx)
+	for i := 0; i < int(nlist); i++ {
+		list_size := C.faiss_IndexIVF_get_list_size(idx.idx, C.size_t(i))
+		invlist := make([]int64, list_size)
+		C.faiss_IndexIVF_invlists_get_ids(idx.idx, C.size_t(i), (*C.idx_t)(&invlist[0]))
+		clusterVectorIDMap[int64(i)] = invlist
+	}
+
+	return clusterVectorIDMap, nil
+}
+
+func (idx *faissIndex) SearchSpecifiedClusters(include, eligibleCentroidIDs []int64,
+	minEligibleCentroids int, k int64, x, centroidDis []float32,
+	params json.RawMessage) ([]float32, []int64, error) {
+	// Applies only to IVF indexes.
+	if ivfIdx := C.faiss_IndexIVF_cast(idx.cPtr()); ivfIdx == nil {
+		return nil, nil, nil
+	}
+
+	includeSelector, err := NewIDSelectorBatch(include)
+	if err != nil {
+		return nil, nil, err
+	}
+	defer includeSelector.Delete()
+
+	tempParams := tempSearchParamsIVF{
+		Nlist: len(eligibleCentroidIDs),
+		// Have to override nprobe so that more clusters will be searched for this
+		// query, if required.
+		Nprobe: minEligibleCentroids,
+		// Only consider the vectors eligible to be searched, based on deletions/
+		// filter queries.
+		Nvecs: len(include),
+	}
+
+	tempParamsBytes, err := json.Marshal(tempParams)
```

Comment:
Create a new API creating search params specific to IVF indexes - avoid marshalling temp params and send as is.